Configure signature version and region on upload

hmstepanek opened the pull request:
An aws signature version must be configured when doing server side
encryption.

Since us-west-1 has permanently been redirected to us-west-2 we need
to override the region on upload but the dump must still happen with
endpoint us-west-1.

I have archived the old docker as v1.01 and added this new version as v1.02 and latest.
![image](https://user-images.githubusercontent.com/30059933/45910572-0e3bbf80-bdbf-11e8-9960-ee5e29211a06.png)

Commit:
Configure signature version and region on upload

An aws signature version must be configured when doing server side
encryption.

Change from --quiet to --only-show-errors.
--quiet suppresses everything including errors whereas --only-show-errors
suppresses everything but errors and warnings.

robertknight commented:
Firstly, 👍 for figuring out the problem with signature versions. However I'm not so sure about the need to add the region. I explain more about that below, but can you explain how you tested this?

> Since us-west-1 has permanently been redirected to us-west-2 we need to override the region on upload but the dump must still happen with endpoint us-west-1.

I think there is a misunderstanding here:

1. Our RDS instance, and almost all of our other AWS resources, are in the us-west-1 region (North California). When you log in to https://aws.hypothes.is, you can see the console's domain is https://us-west-1.console.aws.amazon.com.
2. The `hypothesis-backup-pg` bucket is in the us-west-2 region (Oregon) and _always has been_. I don't know for certain why it was put in a different region. Perhaps as an extra safeguard? Screenshot from S3 console:

<img width="847" alt="screenshot 2018-09-22 07 32 52" src="https://user-images.githubusercontent.com/2458/45914239-0705c780-be3a-11e8-9032-3b8540463525.png">

This morning I pulled the latest version of the postgres-s3-backup Docker image that you uploaded and then ran this locally in the playbook repo to create the container but launch a shell instead of running the backup script:

```
docker run --tty --interactive --rm --name postgres-s3-backup --env-file ./secret/conf/postgres-s3-backup_env hypothesis/postgres-s3-backup:latest /bin/sh
```

I was then able to read and write to the AWS bucket using S3 commands (`aws s3 ls $S3_URL`) without specifying the region.

hmstepanek commented:
The message said permanently redirected and it failed to upload for me without specifying the region. You are correct, it does have to do with the bucket. Probably my commit message should mention something about the bucket. All I know is it failed with a Permanent Redirect error and said I had to specify the region in order to upload it. 🤷‍♀️ I got the impression this was a pretty common thing when I googled it. It’s odd it hasn’t failed before now though.

I tested it by pulling the docker container onto mon and running it. (When I was debugging it I also ran it on mon in interactive mode.) 

Hannah

> On Sep 22, 2018, at 12:01 AM, Robert Knight <notifications@github.com> wrote:
> 
> Firstly, 👍 for figuring out the problem with signature versions. However I'm not so sure about the need to add the region. I explain more about that below, but can you explain how you tested this?
> 
> Since us-west-1 has permanently been redirected to us-west-2 we need to override the region on upload but the dump must still happen with endpoint us-west-1.
> 
> I think there is a misunderstanding here:
> 
> Our RDS instance, and almost all of our other AWS resources, are in the us-west-1 region (North California). When you log in to https://aws.hypothes.is, you can see the console's domain is https://us-west-1.console.aws.amazon.com.
> The hypothesis-backup-pg bucket is in the us-west-2 region (Oregon) and always has been. I don't know for certain why it was put in a different region. Perhaps as an extra safeguard? Screenshot from S3 console:
> 
> 
> This morning I pulled the latest version of the postgres-s3-backup Docker image that you uploaded and then ran this locally in the playbook repo to create the container but launch a shell instead of running the backup script:
> 
> docker run --tty --interactive --rm --name postgres-s3-backup --env-file ./secret/conf/postgres-s3-backup_env hypothesis/postgres-s3-backup:latest /bin/sh
> I was then able to read and write to the AWS bucket using S3 commands (aws s3 ls $S3_URL) without specifying the region.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

robertknight reviewed (changes requested):
This morning I created a Docker image with _just_ the signature version change and pushed it to `robertknight/hypothesis-postgres-s3-backup:kms-test` then ran it on mon. It successfully [uploaded a dump](https://s3.console.aws.amazon.com/s3/object/hypothesis-backup-pg/dumps/postgres-2018-09-24-095724.dump?region=us-west-1&tab=overview).

I also found a blog post with some more details on signatures and SSE-KMS which I thought useful to reference. See https://github.com/hypothesis/postgres-s3-backup/blob/149e74766f2c431c0b85804607af99d873b0d1eb/pg-s3-backup#L52

On that basis, I think we should remove the region change from this PR since it ties the script to a hard-coded region.

robertknight commented:
Update: I did a little more experimentation. Using `aws --debug s3 ls $S3_URL` inside the container provides some more insight:

```
2018-09-24 10:21:14,736 - MainThread - botocore.parsers - DEBUG - Response body:
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AuthorizationHeaderMalformed</Code><Message>The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'us-west-2'</Message><Region>us-west-2</Region><RequestId>28F275B3834CCF67</RequestId><HostId>Kzf3pd7jrGVzNVFqg56t4BkIlM4stPfPgxQ0YfuPiHefMrmhGPUhiaO2F+cWUx9dnDyBUVhMiWc=</HostId></Error>
2018-09-24 10:21:14,737 - MainThread - botocore.hooks - DEBUG - Event needs-retry.s3.ListObjectsV2: calling handler <botocore.retryhandler.RetryHandler object at 0x7f9671bc8210>
2018-09-24 10:21:14,737 - MainThread - botocore.retryhandler - DEBUG - No retry needed.
2018-09-24 10:21:14,737 - MainThread - botocore.hooks - DEBUG - Event needs-retry.s3.ListObjectsV2: calling handler <bound method S3RegionRedirector.redirect_from_error of <botocore.utils.S3RegionRedirector object at 0x7f9671bc8250>>
2018-09-24 10:21:14,737 - MainThread - botocore.utils - DEBUG - S3 client configured for region us-east-1 but the bucket hypothesis-backup-pg is in region us-west-2; Please configure the proper region to avoid multiple unnecessary redirects and signing attempts.
2018-09-24 10:21:14,737 - MainThread - botocore.utils - DEBUG - Updating URI from https://s3.amazonaws.com/hypothesis-backup-pg?list-type=2&delimiter=%2F&prefix=dumps&encoding-type=url to https://s3.us-west-2.amazonaws.com/hypothesis-backup-pg?list-type=2&delimiter=%2F&prefix=dumps&encoding-type=url
```

So it looks like the initial HTTP request made _without_ specifying a region _does fail_, but the "aws" CLI tool is automatically catching the error and sending a redirected request to the correct region.

We could continue to simply rely on this behaviour. It would be marginally more efficient to specify the region, but we ought not to hardcode it in the script but rather accept it as an environment variable in that case. Given that most of the time is spent uploading the actual dump, this isn't particularly urgent though.

hmstepanek commented:
Ah, that makes sense. I did run w/o the quiet flag when I was debugging it. I rather fix this now but there's a command for specifying the bucket and it returning the region so we could prob rely on that rather than making another env var so let me look into that.

robertknight commented:
>  I rather fix this now but there's a command for specifying the bucket

Do you mean `aws s3api get-bucket-location`? Out of interest I did try running that command inside the Docker container this morning but got a permissions error.  Do feel free to double-check this but it might be the case that the AWS keys used to run the dump in production have been given minimal privileges which doesn't include this ability.

hmstepanek commented:
Yes I believe that was it. Hmmm...ok well if that's the case we might have to pass it in instead.

robertknight commented:
> Yes I believe that was it. Hmmm...ok well if that's the case we might have to pass it in instead.

As long as `aws s3 cp` is capable of automatically handling the initial error when it tries to upload, we could also just skip dealing with this for now.

Alternatively, if we did add the region, then we'd first need to modify the config file to add a new env var and deploy that, then update the script and deploy that too.

hmstepanek commented:
Ah yes, I see what you are saying now. With the latest version of the cli it does automatically handle that whereas before the --quiet was just suppressing the error. I no longer see the error when I run without the --quiet. That being said, rather than --quiet I think what we actually want is --only-show-errors. See docs [here](https://docs.aws.amazon.com/cli/latest/reference/s3/cp.html?highlight=only-show-errors).

hmstepanek commented:
Ok, I've removed the --region and replaced --quiet with --only-show-errors. I've uploaded this version to the registry as v1.02. I haven't replaced latest with this version yet, once this is approved then I'll do that.

robertknight reviewed (approved):
LGTM. I haven't tested `--only-show-errors` locally but from the docs it does sound like what we want.

hmstepanek commented:
^ I ran the new image a moment ago on mon with the only-show-errors change and it looked good.

hmstepanek commented:
Ok I'm gonna release the v1.02 as latest and merge this.
